### PR TITLE
start of vm work

### PR DESCRIPTION
**Pull request by tballmsft**
See https://github.com/microsoft/jacdac/issues/374

### CODE REVIEW

File: src/vm/environment.ts [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+}
+
+export class VMServiceEnvironment extends JDServiceClient {
+    private _registers: SMap<JDRegister> = {}
+    private _events: SMap<JDEvent> = {}
+    private _serviceSpec: jdspec.ServiceSpec;
+
+    constructor(service: JDService) {
+        super(service)
+        this._serviceSpec = serviceSpecificationFromClassIdentifier(
```
Comment (pelikhan) [inline src/vm/environment.ts:32]: use ``service.specification``

File: src/vm/environment.ts [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+            }
+        }
+    }
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    public writeRegister(regName: string, ev: any) {
+        const jdreg = this._registers[ regName] 
+        if (jdreg) {
+            const fmt = jdreg.specification?.packFormat
+            jdreg.sendSetPackedAsync(fmt, [ev])
```
Comment (pelikhan) [inline src/vm/environment.ts:72]: await?

Comment (pelikhan) [reply src/vm/environment.ts:72]: you'll get into trouble with the accelerometer and so forth. you want ``[ev]`` to be ``PackedValues`` and handle the arrays yourself.

File: src/vm/expr.ts [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+ 
+We will have a small key-value store to keep program state (perhaps we will have the ability to store lists of values as well as basic values) across the handler executions.
+ 
+Commands can talk to JD services (probably via roles), as well as read/write program state, and wait on events/expressions. Any command can be guarded by an expression, for conditional execution.
+ 
+Expressions can be against service registers (as in the test case) and program state.
+ 
+*/
+
+export type SMap<T> = { [v: string]: T }
```
Comment (pelikhan) [inline src/vm/expr.ts:16]: already in `./utils`

File: src/vm/environment.ts (hunk 2 above)
Comment (tballmsft) [reply src/vm/environment.ts:72]: Code example?

**Pull request merged**